bluez/plugins/jolla-wakelock.c

```diff
@@ -0,0 +1,293 @@
+/*
+ *
+ *  BlueZ - Bluetooth protocol stack for Linux
+ *
+ *  Copyright (C) 2015 Jolla Ltd. All rights reserved.
+ *  Contact: Hannu Mallat <hannu.mallat@jollamobile.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+/* Plugin that implements wakelocks. Only one real wakelock is ever
+   used; API can create several "virtual" wakelocks which are mapped
+   on the real one: 
+
+   - no virtual wakelock active -> real wakelock not active
+   - at least one virtual wakelock active -> real wakelock active
+*/
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <string.h>
+#include <errno.h>
+#include <glib.h>
+
+#include "plugin.h"
+#include "log.h"
+#include "wakelock.h"
+
+#define LOCK_FILE	"/sys/power/wake_lock"
+#define UNLOCK_FILE	"/sys/power/wake_unlock"
+#define LOCK_NAME	"bluetoothd"
+#define LOCK_NAME_LEN   10
+
+struct wakelock {
+	gchar *name;
+	gsize acquisitions;
+	gboolean stale;
+};
+
+static GSList *locks = NULL;
+
+static gboolean initialized = FALSE;
+
+static gboolean real_lock_acquired = FALSE;
+
+static int sysfs_write(const gchar *file, const gchar *data_ptr, gsize data_len)
+{
+	int fd = -1;
+	int r;
+
+	fd = open(file, O_WRONLY);
+	if (fd < 0) {
+		error("Failed to open: %s(%d)", strerror(errno), errno);
```

Comment:
Not sure if it matters here, but strerror() is not thread safe. If error() can be assumed to go to glibc formatter, them "%m" could be used instead of "%s" + strerror(errno)